torch/csrc/api/src/optim/adam.cpp

```diff
@@ -12,57 +12,167 @@
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, step);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(Tensor, exp_avg);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(Tensor, exp_avg_sq);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(Tensor, max_exp_avg_sq);
+}
 
 void Adam::step() {
-  for (size_t i = 0; i < parameters_.size(); ++i) {
-    Tensor p = parameters_.at(i);
-    if (!p.grad().defined()) {
-      continue;
-    }
+  for (auto& group : param_groups_) {
+    for (auto& p : group.params()) {
+      if (!p.grad().defined()) {
+        continue;
+      }
+      auto grad = p.grad().data();
+      TORCH_CHECK(!grad.is_sparse(), "Adam does not support sparse gradients"/*, please consider SparseAdam instead*/);
+      auto param_state = state_.find(c10::guts::to_string(p.unsafeGetTensorImpl()));
+      auto& options = static_cast<AdamOptions&>(group.options());
+      // State initialization
+      if(param_state == state_.end()) {
+        auto state = std::make_unique<AdamParamState>();
+        state->step(0);
+        // Exponential moving average of gradient values
+        state->exp_avg(torch::zeros_like(p.data(), MemoryFormat::Preserve));
+        // Exponential moving average of squared gradient values
+        state->exp_avg_sq(torch::zeros_like(p.data(), MemoryFormat::Preserve));
+        if(options.amsgrad()) {
+          // Maintains max of all exp. moving avg. of sq. grad. values
+          state->max_exp_avg_sq(torch::zeros_like(p.data(), MemoryFormat::Preserve));
+        }
+        state_[c10::guts::to_string(p.unsafeGetTensorImpl())] = std::move(state);
+      }
 
-    if (options.weight_decay() > 0) {
-      NoGradGuard guard;
-      p.grad() = p.grad() + options.weight_decay() * p;
-    }
+      auto& state = static_cast<AdamParamState&>(*state_[c10::guts::to_string(p.unsafeGetTensorImpl())]);
+      auto& exp_avg = state.exp_avg();
+      auto& exp_avg_sq = state.exp_avg_sq();
+      auto& max_exp_avg_sq = state.max_exp_avg_sq();
 
-    auto& exp_average = buffer_at(exp_average_buffers, i);
-    auto& exp_average_sq = buffer_at(exp_average_sq_buffers, i);
-
-    buffer_at(step_buffers, i) += 1;
-    const auto bias_correction1 =
-        1 - std::pow(options.beta1(), buffer_at(step_buffers, i));
-    const auto bias_correction2 =
-        1 - std::pow(options.beta2(), buffer_at(step_buffers, i));
-
-    exp_average.mul_(options.beta1()).add_(p.grad(), 1 - options.beta1());
-    exp_average_sq.mul_(options.beta2())
-        .addcmul_(p.grad(), p.grad(), 1 - options.beta2());
-
-    Tensor denom;
-    if (options.amsgrad()) {
-      auto& max_exp_average_sq = buffer_at(max_exp_average_sq_buffers, i);
-      max_exp_average_sq = torch::max(max_exp_average_sq, exp_average_sq);
-      denom = max_exp_average_sq / bias_correction2;
-    } else {
-      denom = exp_average_sq / bias_correction2;
-    }
+      state.step(state.step()+1);
+      auto beta1 = std::get<0>(options.betas());
+      auto beta2 = std::get<0>(options.betas());
```

Comment:
I think this should be
```suggestion
      auto beta2 = std::get<1>(options.betas());
```